### PR TITLE
[Feature] Add QMS Bing Maps

### PR DESCRIPTION
**Pull request by mind84**
**TL;DR**

Enabling Bing Maps to be loaded as baselayers in LWC through OpenLayers dedicated implementation by importing Bing Maps in QGIS project from [QuickMapServices QGIS Plugin](https://plugins.qgis.org/plugins/quick_map_services/)

**DETAILS**

In previous versions of LWC (i.e. 3.6) it was possible to add a defined set of external layers (including Bing) by flagging the corresponding option in the plugin's `base-layer` panel and without adding any layer to the project. Then the plugin updates the `cfg` accordingly by setting the corresponding properties in the option section of the cfg:

```json
"options": {


        "osmMapnik": "True",
        "openTopoMap": true,
        "bingKey": "some___key",
        "bingStreets": "True",
        "bingSatellite": "True",

}
```
 
Now, since the direction is to rely on the QGIS project, in newer versions this panel has become deprecated and the above properties that defines these external layers are no longer written in `cfg` file.
Furthermore, following the update of the OpenLayers library, some of this external layers are not longer supported (e.g. Google Maps) by the lib itself.

In the "Bing" specific case, the current version of OpenLayers offers supports for rendering Bing tiles.

This implementation allows LWC to recognize the `QMS Bing Maps` from QGIS project, read the [corresponding](https://github.com/3liz/lizmap-plugin/pull/550) `Bing Api Key` from the `cfg` and instantiate with these options the Bing OpenLayers [class](https://openlayers.org/en/latest/apidoc/module-ol_source_BingMaps-BingMaps.html) 

**NOTES**

- This functionality relies for now on the external `url` assigned to the layer by QMS when it is added on QGIS project. This url is set in the `url` property of the `externalAccess` object in the `cfg` response.

- I guess [this object](https://github.com/3liz/lizmap-web-client/blob/2f8d01ce0b67bf1cb24cb8f4446fed9088b05c76/assets/src/modules/config/BaseLayer.js#L751) was only kept for backwards compatibility since the properties listed are no longer written to the `cfg` file. 

- I've extended [this default configuration](https://github.com/3liz/lizmap-web-client/blob/2f8d01ce0b67bf1cb24cb8f4446fed9088b05c76/assets/src/modules/config/BaseLayer.js#L542) with a new `QMSExternalLayer object`, to keep things clearly separated (and hopefully maintainable). 

-  urls of Bing maps currently supported:
   * `Bing Map`, alias `bing-road` in the LWC < 3.7 configuration. url = `https://ecn.dynamic.t0.tiles.virtualearth.net/comp/CompositionHandler/{q}?mkt=en-us&it=G,VE,BX,L,LA&shading=hill`
   * `Bing Satellite`, alias `bing-aerial` in the LWC < 3.7 configuration. url = `https://ecn.t3.tiles.virtualearth.net/tiles/a{q}.jpeg?g=0&dir=dir_n`. 
 
You can check this functionality in the `bing_basemap.qgs` project included in this PR.

Thanks

Related to https://github.com/3liz/lizmap-web-client/issues/4030

Funded by Faunalia

### CODE REVIEW

**Comment by Gustry** [comment]
Thanks for tackling this !
I will let others commenting about the technical part. I thought some work was done about these layers.

Just a note for now :

> This functionality relies for now on the default layer name assigned by QMS when it is added on QGIS project, so to get it work the layer name should not be changed (I'm not sure how limiting this might be)

> Default names of Bing maps currently supported:

This is fragile and will be broken IMHO. (or a dev in the plugin must be done, but I don't think it's the way to go)

In the QGS file of your PR : 

```xml
<layer-tree-layer source="type=xyz&amp;zmin=1&amp;zmax=18&amp;url=https://ecn.t3.tiles.virtualearth.net/tiles/a{q}.jpeg?g%3D0%26dir%3Ddir_n'" checked="Qt::Checked" patch_size="-1,-1" name="Bing Satellite" providerKey="wms" legend_exp="" expanded="1" legend_split_behavior="0" id="Bing_Satellite_9ceaa33e_9c61_4fd0_b607_68ffd5614d88">
```

QGIS is checking if the `{q}` is in the datasource : https://github.com/qgis/QGIS/blob/61fe17774e1eae721b6a616a2bbe934ca399a6ca/src/providers/wms/qgswmsprovider.cpp#L1679
if we want less fragile, we could check the URL is `virtualearth.net` ?

**Comment by mind84** [comment]
> This is fragile and will be broken IMHO. (or a dev in the plugin must be done, but I don't think it's the way to go)

I agree with you and I think this is the most tricky part. Indeed, this is the main reason I've created a draft :)

> if we want less fragile, we could check the URL is virtualearth.net ?

Thanks for checking this! Surely the `url` is much more stable than the `default name`. I will try with this solution but I think that probably it requires some server side implementation.

Thanks

**Comment by mind84** [comment]
@Gustry,

I've changed the logic using the `url` instead of the `name`. It's a little tricky because I've not found a very strong way to separate `roads` from `satellite` map type using just the url.
I don't know if this result would be solid enough/acceptable. This way, however, users will gain more flexibility on the layer name and will be able to change it in QGIS project.

**Review by Gustry — APPROVED** [review]
LGTM from the user point of view

The URL checks looks better like this.

File: tests/js-units/data/bing_basemap-config.json [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        "maxScale": 1000000000,
+        "use_native_zoom_levels": true,
+        "hide_numeric_scale_value": false,
+        "initialExtent": [
+            398421.7687,
+            5386390.8765,
+            467349.1746,
+            5427483.4229
+        ],
+        "bingKey": "AgBOKzHM1S17uF6szXY93vepXIx2Cl91mkrlP-yrsngz1jdREjrtucUduz56hDNT",
```
Comment (nworr) [inline tests/js-units/data/bing_basemap-config.json:179]: is this key a dev key ? can it be exposed publicly  ?

File: tests/qgis-projects/tests/bing_basemap.qgs.cfg [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        "maxScale": 1000000000,
+        "use_native_zoom_levels": true,
+        "hide_numeric_scale_value": false,
+        "initialExtent": [
+            398421.7687,
+            5386390.8765,
+            467349.1746,
+            5427483.4229
+        ],
+        "bingKey": "AgBOKzHM1S17uF6szXY93vepXIx2Cl91mkrlP-yrsngz1jdREjrtucUduz56hDNT",
```
Comment (nworr) [inline tests/qgis-projects/tests/bing_basemap.qgs.cfg:37]: same question

**Review by nworr — COMMENTED** [review]
Good PR, thanks for the detailed description. 
Only  a question about the exposed bing key

File: tests/js-units/node/state/baselayer.test.js [hunk 3]
```diff
@@ -290,4 +290,26 @@ describe('BaseLayersState', function () {
...
+    it('Bing Maps baselayers', function () {
+        const baseLayers = getBaseLayersState('bing_basemap')
+        expect(baseLayers.selectedBaseLayerName).to.be.eq('bing roads')
+        expect(baseLayers.selectedBaseLayer).to.not.be.undefined
+        expect(baseLayers.selectedBaseLayer.name).to.be.eq('bing roads')
+        expect(baseLayers.baseLayerNames).to.be.an('array')
+        .that.have.length(2)
+        .that.be.deep.eq([
+            "bing roads",
+            "bing areals",
```
Comment (Gustry) [inline tests/js-units/node/state/baselayer.test.js:302]:
```suggestion
            "bing aerial",
```
?

File: tests/qgis-projects/tests/bing_basemap.qgs [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+        <Option type="Map">
+          <Option type="QString" value="baselayers" name="wmsShortName"/>
+        </Option>
+      </customproperties>
+      <layer-tree-layer legend_split_behavior="0" expanded="1" providerKey="wms" source="crs=EPSG:3857&amp;format&amp;type=xyz&amp;url=https://ecn.dynamic.t0.tiles.virtualearth.net/comp/CompositionHandler/%7Bq%7D?mkt%3Den-us%26it%3DG,VE,BX,L,LA%26shading%3Dhill&amp;zmax=19&amp;zmin=1" id="Bing_Map_a346c65b_6c10_4b31_80b7_9ad63af606b6" name="bing roads" checked="Qt::Checked" legend_exp="" patch_size="-1,-1">
+        <customproperties>
+          <Option/>
+        </customproperties>
+      </layer-tree-layer>
+      <layer-tree-layer legend_split_behavior="0" expanded="1" providerKey="wms" source="crs=EPSG:3857&amp;format&amp;type=xyz&amp;url=https://ecn.t3.tiles.virtualearth.net/tiles/a%7Bq%7D.jpeg?g%3D0%26dir%3Ddir_n'&amp;zmax=18&amp;zmin=1" id="Bing_Satellite_9ceaa33e_9c61_4fd0_b607_68ffd5614d88" name="bing areals" checked="Qt::Unchecked" legend_exp="" patch_size="-1,-1">
```
Comment (Gustry) [inline tests/qgis-projects/tests/bing_basemap.qgs:35]:
```suggestion
      <layer-tree-layer legend_split_behavior="0" expanded="1" providerKey="wms" source="crs=EPSG:3857&amp;format&amp;type=xyz&amp;url=https://ecn.t3.tiles.virtualearth.net/tiles/a%7Bq%7D.jpeg?g%3D0%26dir%3Ddir_n'&amp;zmax=18&amp;zmin=1" id="Bing_Satellite_9ceaa33e_9c61_4fd0_b607_68ffd5614d88" name="bing aerial" checked="Qt::Unchecked" legend_exp="" patch_size="-1,-1">
```

File: tests/qgis-projects/tests/bing_basemap.qgs [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+  </mapcanvas>
+  <projectModels/>
+  <legend updateDrawingOrder="true">
+    <legendgroup open="true" name="baselayers" checked="Qt::Checked">
+      <legendlayer showFeatureCount="0" open="true" drawingOrder="-1" checked="Qt::Checked" name="bing roads">
+        <filegroup open="true" hidden="false">
+          <legendlayerfile layerid="Bing_Map_a346c65b_6c10_4b31_80b7_9ad63af606b6" isInOverview="0" visible="1"/>
+        </filegroup>
+      </legendlayer>
+      <legendlayer showFeatureCount="0" open="true" drawingOrder="-1" checked="Qt::Unchecked" name="bing areals">
```
Comment (Gustry) [inline tests/qgis-projects/tests/bing_basemap.qgs:84]:
```suggestion
      <legendlayer showFeatureCount="0" open="true" drawingOrder="-1" checked="Qt::Unchecked" name="bing aerial">
```

File: tests/qgis-projects/tests/bing_basemap.qgs [hunk 6]
```diff
@@ -0,0 +1,778 @@
...
+        <xmax>179.99999999999997158</xmax>
+        <ymax>85.05112877980660357</ymax>
+      </wgs84extent>
+      <id>Bing_Satellite_9ceaa33e_9c61_4fd0_b607_68ffd5614d88</id>
+      <datasource>crs=EPSG:3857&amp;format&amp;type=xyz&amp;url=https://ecn.t3.tiles.virtualearth.net/tiles/a%7Bq%7D.jpeg?g%3D0%26dir%3Ddir_n'&amp;zmax=18&amp;zmin=1</datasource>
+      <shortname>Bing_Satellite</shortname>
+      <keywordList>
+        <value></value>
+      </keywordList>
+      <layername>bing areals</layername>
```
Comment (Gustry) [inline tests/qgis-projects/tests/bing_basemap.qgs:365]:
```suggestion
      <layername>bing aerial</layername>
```

File: tests/qgis-projects/tests/bing_basemap.qgs.cfg [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            "groupAsLayer": "False",
+            "baseLayer": "False",
+            "displayInLegend": "True",
+            "group_visibility": [],
+            "singleTile": "True",
+            "imageFormat": "image/png",
+            "cached": "False",
+            "clientCacheExpiration": 300
+        },
+        "bing areals": {
```
Comment (Gustry) [inline tests/qgis-projects/tests/bing_basemap.qgs.cfg:110]: same here and other lines below

**Review by Gustry — COMMENTED** [review]
Is-it a typo ?

File: tests/js-units/data/bing_basemap-config.json (hunk 1 above)
Comment (mind84) [reply tests/js-units/data/bing_basemap-config.json:179]: Hi @nworr,
I thought I could used it because it is already [exposed](https://github.com/3liz/lizmap-web-client/blob/2d836511de728cf24730a53ee98b0febbbd32ebb/tests/qgis-projects/tests/base_layers.qgs.cfg#L45) in `base_layers.qgs.cfg`. 

Thanks

File: tests/js-units/node/state/baselayer.test.js (hunk 3 above)
Comment (mind84) [reply tests/js-units/node/state/baselayer.test.js:302]: It is

File: tests/js-units/data/bing_basemap-config.json (hunk 1 above)
Comment (nworr) [reply tests/js-units/data/bing_basemap-config.json:179]: i didn't even notice this project use bing map :)

**Review by nworr — APPROVED** [review]

**Comment by Gustry** [comment]
Just for information, see #5944
